sensor: icm42688: Fix broken streaming, add tests and enable in sensor_shell sample.

bperseghetti opened the pull request:
This fixes the broken streaming values we were getting from icm42688 in high res and normal modes.

Also adds a test to make sure the values never blow up again by using real hardware (including the vmu_rt1176). 

fixes #84833

teburd commented on drivers/sensor/tdk/icm42688/icm42688_common.c:
```diff
@@ -255,8 +255,11 @@ int icm42688_configure(const struct device *dev, struct icm42688_cfg *cfg)
+			FIELD_PREP(BIT_FIFO_GYRO_EN, 1) |
+			FIELD_PREP(BIT_FIFO_ACCEL_EN, 1) |
+			FIELD_PREP(BIT_FIFO_TMST_FSYNC_EN, 1) |
+			FIELD_PREP(BIT_FIFO_HIRES_EN, 1);
```
hires should probably be a config option, would be fine to set this in DT statically imho.

bperseghetti replied:
It is now.

teburd commented on drivers/sensor/tdk/icm42688/icm42688_decoder.c:
```diff
@@ -234,81 +234,33 @@ static inline q31_t icm42688_read_temperature_from_packet(const uint8_t *pkt)
-	}
+	const uint32_t scale[2][2] = {
+		/* low-res,	hi-res */
+		{35744,		8936}, /* gyro */
```
Having the scales precalculated in a table is great, but at least a formula for how these are calculated should be in a comment here

teburd replied:
Really, I think I think a macro could be useful here for computing a scaling factor

E.g.

SENSOR_Q31_SCALER(sensitivity, shift) is half the battle, e.g. getting the math for most sensor conversions looking like the simple integer multiplier math that it should be, could be done simpler I think in most cases. Better still I love const table here, though sadly C doesn't let us do const math expressions like the following even if the compiler will do it at compile time.

```c
const sensitivity = SENSITIVITIES[drv_cfg->fs];
const shift = SHIFTS[drv_cfg->fs];
const int32_t scaler = (1/sensitivity)*(pow(2,31)/pow(2,shift));
q31_t scaled = sample*scaler;
```

There's also occasionally the unit conversion involved, e.g. converting G to m/s^2 would involve another multiplier of 9.81 m/s^2. But clearly it was wrong here and needed to be rectified.

teburd commented on drivers/sensor/tdk/icm42688/icm42688_decoder.c:
```diff
@@ -234,81 +234,33 @@ static inline q31_t icm42688_read_temperature_from_packet(const uint8_t *pkt)
 	}
 
-	*out = (q31_t)(value * scale / max);
+	*out = (q31_t)(signed_value * scale[is_accel][is_hires]);
```
I like that a lot of the math here comes down to a simple multiply of a constant. That's actually quite nice.

The constants do need to be documented as to how they are calculated though, or better be calculated at compile time by the compiler.

teburd reviewed (commented):
The test needs a testcase.yaml and likely a fixture assigned to it

bperseghetti committed:
sensor: icm42688: Fix broken build and streaming.

Makes driver actually build again.
Fixes scaling, q32 encoding, bit shifts, and register mappings.
Also enables High Resolution mode (a unique and beneficial
feature of the chip when used with streaming).

Co-authored-by: James Goppert <james.goppert@gmail.com>
Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>

bperseghetti committed:
tests: boards: vmu_rt1170: icm42688: hardware test

Adds a test for icm42688 on vmu_rt1170.

Co-authored-by: James Goppert <james.goppert@gmail.com>
Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>

bperseghetti committed:
samples: sensor: sensor_shell: add vmu_rt1170.

Adds support for testing sensor shell streaming
on the vmu_rt1170 with RTIO.

Co-Authored-by: James Goppert <james.goppert@gmail.com>
Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>

bperseghetti commented:
After this gets merged I plan to open a PR for some more features that we have been using regularly they are in this post-merge branch, and here is what it adds (note I will split it out into better commits). [This is the planned post-merge additions as a compare.](https://github.com/CogniPilot/zephyr/compare/pr-fix-icm42688-stream...CogniPilot:zephyr:pr-post-merge-icm42688-stream-fix?expand=1)

teburd reviewed (approved):
Approving as its fixing the driver.

Realistically the scale values are under-documented and specific to a particular predefined accel/gyro scale which isn't quite right. It should be a table of scaling factors dependent on the sensor scale (e.g. 2g, 4g, 16g, etc). It would be great to see this fixed in a follow up.
